ouroboros-consensus-cardano/src/ouroboros-consensus-cardano/Ouroboros/Consensus/Cardano/Ledger.hs

```diff
@@ -0,0 +1,209 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Ouroboros.Consensus.Cardano.Ledger (CardanoTxOut (..)) where
+
+import qualified Cardano.Ledger.Core as Core
+import qualified Cardano.Ledger.Shelley.API as SL
+import qualified Codec.CBOR.Decoding as CBOR
+import qualified Codec.CBOR.Encoding as CBOR
+import           Data.SOP.Index
+import qualified Data.SOP.InPairs as InPairs
+import           Data.Void
+import           GHC.Generics
+import           NoThunks.Class
+import           Ouroboros.Consensus.Cardano.Block
+import           Ouroboros.Consensus.Cardano.CanHardFork
+import           Ouroboros.Consensus.HardFork.Combinator
+import           Ouroboros.Consensus.HardFork.Combinator.State.Types
+import           Ouroboros.Consensus.Ledger.Tables
+import           Ouroboros.Consensus.Protocol.Praos (Praos)
+import           Ouroboros.Consensus.Protocol.TPraos (TPraos)
+import           Ouroboros.Consensus.Shelley.Ledger (ShelleyBlock)
+
+instance CardanoHardForkConstraints c
+      => HasCanonicalTxIn (CardanoEras c) where
+  newtype instance CanonicalTxIn (CardanoEras c) = CardanoTxIn {
+      getCardanoTxIn :: SL.TxIn c
+    }
+    deriving stock (Show, Eq, Ord)
+    deriving newtype NoThunks
+
+  injectCanonicalTxIn IZ       byronTxIn   = absurd byronTxIn
+  injectCanonicalTxIn (IS idx) shelleyTxIn = case idx of
+      IZ                               -> CardanoTxIn shelleyTxIn
+      IS IZ                            -> CardanoTxIn shelleyTxIn
+      IS (IS IZ)                       -> CardanoTxIn shelleyTxIn
+      IS (IS (IS IZ))                  -> CardanoTxIn shelleyTxIn
+      IS (IS (IS (IS IZ)))             -> CardanoTxIn shelleyTxIn
+      IS (IS (IS (IS (IS IZ))))        -> CardanoTxIn shelleyTxIn
+      IS (IS (IS (IS (IS (IS idx'))))) -> case idx' of {}
+
+  distribCanonicalTxIn IZ _                 =
+      error "distribCanonicalTxIn: Byron has no TxIns"
+  distribCanonicalTxIn (IS idx) cardanoTxIn = case idx of
+      IZ                               -> getCardanoTxIn cardanoTxIn
+      IS IZ                            -> getCardanoTxIn cardanoTxIn
+      IS (IS IZ)                       -> getCardanoTxIn cardanoTxIn
+      IS (IS (IS IZ))                  -> getCardanoTxIn cardanoTxIn
+      IS (IS (IS (IS IZ)))             -> getCardanoTxIn cardanoTxIn
+      IS (IS (IS (IS (IS IZ))))        -> getCardanoTxIn cardanoTxIn
+      IS (IS (IS (IS (IS (IS idx'))))) -> case idx' of {}
+
+  encodeCanonicalTxIn   = Core.toEraCBOR @(ShelleyEra c) . getCardanoTxIn
+
+  decodeCanonicalTxIn = CardanoTxIn <$> Core.fromEraCBOR @(ShelleyEra c)
+
+-- Unpacking the fields of the era-specific TxOuts could save a chunk of memory.
+-- However, unpacking of sum types is only possible on @ghc-9.6.1@ and later, so
+-- before @ghc-9.6.1@ we only unpack the TxOuts for eras before Alonzo.
```

Comment:
The ledger's TxOut for eras before Alonzo are product types, and since Alonzo they are sum types. Since `ghc-9.6.1`, we have support for unpacking sum types, but not for older versions. So, we pay the cost of an extra indirection on older ghcs

Reply:
Ahh, I misinterpreted/misread the text somehow, and thought it meant that unpacking a field in a constructor of a sum type is only possible since 9.6.1. However, it actually is a statement about the whether the field is a sum type. Thanks!